aws/resource_aws_datasync_agent.go

```diff
@@ -150,7 +150,7 @@ func resourceAwsDataSyncAgentCreate(d *schema.ResourceData, meta interface{}) er
 	err = resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
 		_, err := conn.DescribeAgent(descAgentInput)
 
-		if isAWSErr(err, "InvalidRequestException", "not found") {
+		if isAWSErr(err, "InvalidRequestException", "does not exist") {
```

Comment:
Would it make sense to encapsulate this check in a function, rather than doing the string comparison in three places in this file and two more in the tests?

Reply:
These `isAWSErr()` checks do tend to need to be copy-pasted like this for every resource (in the `Read` to trigger recreation, `Delete` to ignore on deletion, and `testAccCheck...Destroy()` testing).

It might be nice to think about this problem more globally. The flip side is not trying to introduce abstractions for minor gains or introduce this abstraction in just one or few resources so it becomes inconsistent with the wider codebase.

The quickest abstraction would be something like:

```go
func isDatasyncAgentNotFoundError(err error) bool {
  return isAWSErr(err, "InvalidRequestException", "does not exist")
}
```

But that doesn't feel like we're gaining much. Another (arguably better) abstraction would be if we could define each resource with a mapping of "not found" error codes and messages since that is the crux of what we want to declare:

```go
&AwsResource{ // new wrapper type that extends schema.Resource
  NotFoundErrors: map[string]string{
    "InvalidRequestException": "does not exist",
  },
  Create: // ...
}
```

This would play well with resources that have multiple error conditionals like association resources, e.g. `aws_ec2_client_vpn_network_association`:

```go
&AwsResource{ // new wrapper type that extends schema.Resource
  NotFoundErrors: map[string]string{
    "InvalidClientVpnAssociationId.NotFound": "",
    "InvalidClientVpnEndpointId.NotFound", "",
  },
  Create: // ...
}
```

Then we'd need some function to extract that from the schema and loop through those:

```go
func (r *AwsResource) isNotFoundError(err error) bool {
  for code, message := range r.NotFoundErrors {
    if isAWSErr(err, code, message) {
      return true
    }
  }

  return false
}

// usage
_, err := conn.DoSomething(input)

if resourceAwsDataSyncAgent().isNotFoundError(err) {
  // guessing this could be done better
}
```

Any which way, we'd want to ensure it either compliments/enhances the existing Terraform Plugin SDK `resource.NotFoundError` / `isResourceNotFoundError()` or is not named similarly to prevent confusion.

Reply:
@gdavison what would you like me to do here?

Reply:
Yeah, I was thinking of the first case. More of an encapsulation of the logic, rather than anything more general. I'm good either way.